pkg/compiler/global_test.go

```diff
@@ -80,6 +81,536 @@ func TestUnusedGlobal(t *testing.T) {
+			opcode.PUSH1, opcode.RET, // Main
+			opcode.PUSH2, opcode.RET) // f
+	})
+	t.Run("used in unused function", func(t *testing.T) {
+		src := `package foo
+			var Unused1, Unused2, Unused3 = 1, 2, 3
+			func Main() int {
+				return 1
+			}
+			func unused1() int {
+				return Unused1
+			}
+			func unused2() int {
+				return Unused1 + unused1()
+			}
+			func unused3() int {
+				return Unused2 + unused2()
+			}`
+		prog := eval(t, src, big.NewInt(1))
+		require.Equal(t, 2, len(prog)) // Main
+	})
+	t.Run("used in used function", func(t *testing.T) {
+		src := `package foo
+			var A = 1
+			func Main() int {
+				return f()
+			}
+			func f() int {
+				return A
+			}`
+		eval(t, src, big.NewInt(1))
+		checkInstrCount(t, src, 1, 1, 0, 0)
+	})
+	t.Run("unused, initialized via init", func(t *testing.T) {
+		src := `package foo
+			var A int
+			func Main() int {
+				return 2
+			}
+			func init() {
+				A = 1		// Although A is unused from exported functions, it's used from init(), so it should be mark as "used" and stored.
+			}`
+		eval(t, src, big.NewInt(2))
+		checkInstrCount(t, src, 1, 0, 0, 0)
+	})
+	t.Run("used, initialized via init", func(t *testing.T) {
+		src := `package foo
+			var A int
+			func Main() int {
+				return A
+			}
+			func init() {
+				A = 1
+			}`
+		eval(t, src, big.NewInt(1))
+		checkInstrCount(t, src, 1, 0, 0, 0)
+	})
+	t.Run("unused, initialized by function call", func(t *testing.T) {
+		t.Run("unnamed", func(t *testing.T) {
+			src := `package foo
+			var _ = f()
+			func Main() int {
+				return 1
+			}
+			func f() int {
+				return 2
+			}`
+			eval(t, src, big.NewInt(1))
+			checkInstrCount(t, src, 0, 1, 0, 0)
+		})
+		t.Run("named", func(t *testing.T) {
+			src := `package foo
+			var A = f()
+			func Main() int {
+				return 1
+			}
+			func f() int {
+				return 2
+			}`
+			eval(t, src, big.NewInt(1))
+			checkInstrCount(t, src, 0, 1, 0, 0)
+		})
+		t.Run("named, with dependency on unused var", func(t *testing.T) {
+			t.Run("drop value", func(t *testing.T) {
+				src := `package foo
+			var (
+				A = 1
+				B = A + 1 // To check nested ident values.
+				C = 3
+				D = B + f() + C // To check that both idents (before and after the call to f) will be marked as "used".
+			)
+			func Main() int {
+				return 1
+			}
+			func f() int {
+				return 2
+			}`
+				eval(t, src, big.NewInt(1), []interface{}{opcode.INITSSLOT, []byte{3}}, // sslot for A
+					opcode.PUSH1, opcode.STSFLD0, // store A
+					opcode.LDSFLD0, opcode.PUSH1, opcode.ADD, opcode.STSFLD1, // store B
+					opcode.PUSH3, opcode.STSFLD2, // store C
+					opcode.LDSFLD1, []interface{}{opcode.CALL, []byte{9}}, opcode.ADD, opcode.LDSFLD2, opcode.ADD, opcode.DROP, opcode.RET, // evaluate D and drop
+					opcode.PUSH1, opcode.RET, // Main
+					opcode.PUSH2, opcode.RET) // f
+			})
+			t.Run("drop value", func(t *testing.T) {
```

Comment:
The same test name as above.

Reply:
I've squashed these two test cases, fixed.